war/src/main/java/io/lumeer/engine/controller/LinkingFacade.java

```diff
@@ -122,52 +92,61 @@ public void onDropDocument(@Observes(notifyObserver = Reception.IF_EXISTS) final
+            String toDocumentId = linkDirection == Linking.LinkDirection.FROM ? doc.getString(Linking.LinkingTable.ATTR_TO_ID) : doc.getString(Linking.LinkingTable.ATTR_FROM_ID);
+            links.add(new LinkDao(fromCollection, toCollection, attrRole, fromDocumentId, toDocumentId, doc.getDataDocument(Linking.LinkingTable.ATTR_ATTRIBUTES)));
+         }
       }
       return links;
    }
 
    /**
     * Read all links between two documents
     *
-    * @param fromCollectionName
+    * @param firstCollectionName
     *       the name of the first document's collection
-    * @param fromId
+    * @param firstDocumentId
     *       the id of the first document
-    * @param toCollectionName
+    * @param secondCollectionName
     *       the name of the second document's collection
-    * @param toId
+    * @param secondDocumentId
     *       the id of the second document
     * @param role
     *       role name
     * @param linkDirection
     *       direction of link
     * @return list of all links
-    * @throws DbException
-    *       When there is an error working with the database.
     */
-   public List<LinkDao> readDocByDocLinks(final String fromCollectionName, final String fromId, final String toCollectionName, final String toId, final String role, final LumeerConst.Linking.LinkDirection linkDirection) throws DbException {
+   public List<LinkDao> readDocByDocLinks(final String firstCollectionName, final String firstDocumentId, final String secondCollectionName, final String secondDocumentId, final String role, final Linking.LinkDirection linkDirection) {
+      List<DataDocument> linkingDocs = readLinkingDocsFromTo(firstCollectionName, secondCollectionName, role, linkDirection);
+
       List<LinkDao> links = new ArrayList<>();
-      List<DataDocument> linkingTables = readLinkingTablesFromTo(fromCollectionName, toCollectionName, role, linkDirection);
-
-      for (DataDocument lt : linkingTables) {
-         String colName = lt.getString(LumeerConst.Linking.MainTable.ATTR_COL_NAME);
-         String fromCollection = lt.getString(LumeerConst.Linking.MainTable.ATTR_FROM_COLLECTION);
-         String toCollection = lt.getString(LumeerConst.Linking.MainTable.ATTR_TO_COLLECTION);
-         List<DataDocument> ls = readLinkingDocumentsFromTo(colName, fromId, toId, linkDirection);
-         links.addAll(convertLinkDaosFromDocuments(ls, fromCollection, toCollection, role));
+      String collectionName = buildCollectionName();
+      for (DataDocument lt : linkingDocs) {
+         String fromCollection = lt.getString(Linking.MainTable.ATTR_FROM_COLLECTION);
+         String toCollection = lt.getString(Linking.MainTable.ATTR_TO_COLLECTION);
+         String attrRole = lt.getString(Linking.MainTable.ATTR_ROLE);
+
+         DataDocument dc = dataStorage.readDocument(collectionName, linkingDocumentsFilter(lt.getId(), firstDocumentId, secondDocumentId, linkDirection));
```

Comment:
We can't, because for each `LinkType` between two collections there is unique id to use as filter in `LinkInstance` collection.